src/scheduler/AsyncMsgNotifyBasedScheduler.cpp

```diff
@@ -177,15 +202,23 @@ void AsyncMsgNotifyBasedScheduler::runLoop(std::vector<folly::Future<Status>>&&
                                             LoopExecutor* loop,
                                             folly::Executor* runner,
                                             std::vector<folly::Promise<Status>>&& promises) const {
-    folly::collect(futures).via(runner).thenValue(
-        [loop, runner, pros = std::move(promises), this](std::vector<Status>&& status) mutable {
+    folly::collect(futures).via(runner).thenTry(
+        [loop, runner, pros = std::move(promises), this](auto&& t) mutable {
+            if (t.hasException()) {
+                return notifyError(pros, Status::Error(t.exception().what()));
+            }
+            auto status = std::move(t).value();
             auto s = checkStatus(std::move(status));
             if (!s.ok()) {
                 return notifyError(pros, s);
             }
 
-            std::move(execute(loop)).thenValue(
-                [loop, runner, pros = std::move(pros), this](Status loopStatus) mutable {
+            std::move(execute(loop)).thenTry(
```

Comment:
I think you could try the `thenError` method of future.

Reply:
The `promise` will no longer valid after `thenTry`, the `thenError` is not an option.